compiler/vm/vmdef.nim

```diff
@@ -90,27 +90,138 @@ type
-    of rkRegisterAddr: regAddr*: ptr TFullReg
-    of rkNodeAddr: nodeAddr*: ptr PNode
-
+    slotEmpty,        ## slot is unused
+    slotFixedVar,     ## slot is used for a fixed var/result (requires copy then)
+    slotFixedLet,     ## slot is used for a fixed param/let
+    slotTempUnknown,  ## slot but type unknown (argument of proc call)
+    slotTempInt,      ## some temporary int
+    slotTempFloat,    ## some temporary float
+    slotTempStr,      ## some temporary string
+    slotTempComplex,  ## some complex temporary (s.node field is used)
+    slotTempPerm      ## slot is temporary but permanent (hack)
+
+  AtomKind* = enum
+    akInt
+    akFloat
+    akString # TODO: maybe even collapse akString into akOrdinalArray?
+    akOrdinalArray # Space efficient storing of float/int arrays
+    akArray # Used for both seqs and fixed size arrays by vmgen
+    akSet
+    akRef
+    akPtr
+    akPointer ## A raw pointer points to inside akString's string or akOrdinalArray/akArray's rawArrayVal
+    akObject
+    akPNode
+    akCallable
+
+  CallableKind* = enum
+    ckDefault ## A referenc to a VM function
+    ckClosure
+    ckCallback ## A VmCallback
+    ckFFI
+    ckTemplate
+
+  # XXX: design is temporary
+  VmFunctionObject* = object ## An object storing a callable
+    case kind*: CallableKind
+    of ckDefault:
+      prcSym*: PNode # TODO: needs to be a `PSym`
+    of ckClosure:
+      closureEnv*: GlobalRef ## A reference to the closure environment
+      closureTarget*: int
+    of ckCallback:
+      cbOffset*: int
+    of ckFFI:
+      ffiPrcSym*: PSym
+    of ckTemplate:
+      templateSym*: PSym
+
+
+  OrdinalArrayKind* = enum
+    oakUnsigned
+    oakSigned
+    oakFloat32
+    oakFloat
+
+  VmObjectAtom* = object
+    case kind*: AtomKind
+    of akInt:
+      # XXX: is it okay to use this for unsigned integer too?
+      intVal*: BiggestInt
+    of akFloat:
+      floatVal*: BiggestFloat
+    of akString:
+      strVal*: string
+    of akOrdinalArray:
+      rawArrayVal*: seq[byte]
+      elemSize*: int8 # 1, 2, 4 or 8
+      oaKind*: OrdinalArrayKind # TODO: encode `oaKind` and `elemTyp` into a single int8
+    of akArray:
+      arrayElements*: seq[VmObjectAtom]
+      elemTyp*: PType
+    of akSet:
+      # Theoretically sets could use nim `set`s for representing data. But that would more
+      # or less require the use of an heap-allocated structure to contain the set and also casing
+      # between set[uint8] and set[uint16] since otherwise `VmObjectAtom` would explode to the size
+      # of 8192 byte (sizeof(set[uint16]))!
+
+      # XXX maybe use `ptr IntSet`? Can't use IntSet with automatic storage duration since
+      # sizeof(IntSet) is also huge (smaller than set[uint16] however)
+
+      # XXX should `setElems` be ordered? Would speed up contains checks but require sorting on modification
+      # XXX `setElem` could also be run-length encoded for speed and memory efficiency...
+      setElems* : seq[uint16]
```

Comment:
Since sets have their size defined at compile time, you might be able to get by with something like a `akOrdinalArray` then reuse the bit set implementation in the stdlib to manipulate/interpret it.